target/ppc/fpu_helper.c

```diff
@@ -3339,3 +3339,198 @@ void helper_xssubqp(CPUPPCState *env, uint32_t opcode,
+        uint8_t xmsk_bit, ymsk_bit, ymax;                               \
+        ppc_vsr_t *a, *b, *at;                                          \
+        float_status *excp_ptr;                                         \
+        TYPE psum, aux_acc, va, vb;                                     \
+        target_ulong fpscr;                                             \
+        int op_flags;                                                   \
+        int i, j;                                                       \
+        bool acc, neg_mul, neg_acc;                                     \
+        xmsk = mask & 0x0F;                                             \
+        ymsk = (mask >> 4) & 0x0F;                                      \
+        ymax = MIN(4, 128/(sizeof(TYPE) * 8));                          \
+        b = cpu_vsr_ptr(env, b_r);                                      \
+        excp_ptr = &env->fp_status;                                     \
+        fpscr = 0;                                                      \
+        acc = ger_acc_flag(packed_flags); /* Accumulate */              \
+        neg_acc = ger_neg_acc_flag(packed_flags); /* Neg Accumulate */  \
+        neg_mul = ger_neg_mul_flag(packed_flags); /* Neg Multiply */    \
+        for (i = 0, xmsk_bit = 1 << 3; i < 4; i++, xmsk_bit >>= 1) {    \
+            a = cpu_vsr_ptr(env, a_r + i/ymax);                         \
+            at = cpu_vsr_ptr(env, at_r + i);                            \
+            for (j = 0, ymsk_bit = 1 << (ymax-1); j < ymax;             \
+                 j++, ymsk_bit >>= 1) {                                 \
+                if ((xmsk_bit & xmsk) && (ymsk_bit & ymsk)) {           \
+                    helper_reset_fpstatus(env);                         \
+                    psum = TYPE##_zero;                                 \
+                    op_flags = (neg_acc ^ neg_mul) ?                    \
+                                          float_muladd_negate_c : 0;    \
+                    op_flags |= (neg_mul) ?                             \
+                                     float_muladd_negate_result : 0;    \
+                    va = a->Vsr##EL(i%ymax);                            \
+                    vb = b->Vsr##EL(j);                                 \
+                    aux_acc = at->Vsr##EL(j);                           \
+                    if (acc) {                                          \
+                        psum = TYPE##_muladd(va, vb, aux_acc,           \
+                                             op_flags, excp_ptr);       \
+                    } else {                                            \
+                        psum = TYPE##_mul(va, vb, excp_ptr);            \
+                    }                                                   \
+                                                                        \
+                    at->Vsr##EL(j) = psum;                              \
```

Comment:
If you use a ternary operator you could even remove `psum` from this function

Reply:
If the objective is removing the psum you don't even need a ternary operator, psum is there mostly to imitate the ISA's pseudocode, but I'll change to a ternary